### ISSUE TITLE
HTTP::Client ssl_verify option

### ISSUE THREAD

**Issue by kostya** [crystal-lang/crystal#2073]
i tried implement it like this https://github.com/kostya/crystal/commit/d2e8a3af533f7e737d6706754e933460977560bc

but something not worked (sites for test: https://badssl.com)
```cr
require "http/client"

client = HTTP::Client.new "expired.badssl.com", nil, true
client.ssl_verify = :none
p client.get "/"

client = HTTP::Client.new "expired.badssl.com", nil, true
client.ssl_verify = :peer
p client.get "/"
```

```
curl https://expired.badssl.com/
curl -k https://expired.badssl.com/
```

any ideas?

**Comment by waj** [comment]
Your implementation looks fine, but still it doesn't work on my machine either. I have no idea what's missing, however some observations:

1. Would the connection fail? Or should we check the result of `SSL_connect` in the SSLSocket?
2. Seems like the SSL layer doesn't know about the host name, so it could not verify the certificate completely. It should be able to verify expired certs though, but dunno if it does any verification at all unless it has all the info?
3. I'd set the verification option just for the socket, using `SSL_set_verify`, instead of changing the default global context that would affect all the connections.

**Comment by mhelmetag** [comment]
Code:
```
require "http/client"

client = HTTP::Client.new("expired.badssl.com", nil, true)
puts client.get("/").inspect
```

Response:
```
#<HTTP::Client::Response:0x1041daac0 @status_code=200, @body="<!DOCTYPE html>\n<html>\n<head>\n  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\n  <link rel=\"shortcut icon\" href=\"/icons/favicon-red.ico\"/>\n  <link rel=\"apple-touch-icon\" href=\"/icons/icon-red.png\"/>\n  <title>expired.badssl.com</title>\n  <link rel=\"stylesheet\" href=\"/style.css\">\n  <style>body { background: red; }</style>\n</head>\n<body>\n<div id=\"content\">\n  <h1 style=\"font-size: 12vw;\">\n    expired.<br>badssl.com\n  </h1>\n</div>\n\n</body>\n</html>\n", @headers=HTTP::Headers{"Server" => "nginx/1.6.2 (Ubuntu)", "Date" => "Wed, 16 Mar 2016 08:00:21 GMT", "Content-Type" => "text/html", "Last-Modified" => "Sat, 19 Dec 2015 02:42:32 GMT", "Transfer-Encoding" => "chunked", "Connection" => "keep-alive", "Cache-Control" => "no-store", "Content-Encoding" => "gzip"}, @version="HTTP/1.1", @body_io=nil, @status_message="OK">
```

I wonder what happened to `client.ssl_verify` since that doesn't seem to be a valid attribute anymore.

**Comment by rubiojr** [comment]
@waj given this, looks like `HTTP::Client` does not verify certs by default:

```crystal
require "http/client"

client = HTTP::Client.new("self-signed.badssl.com", nil, true)
puts client.get("/").body
```

```
$ crystal test.cr
<!DOCTYPE html>
<html>
<head>
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link rel="shortcut icon" href="/icons/favicon-red.ico"/>
  <link rel="apple-touch-icon" href="/icons/icon-red.png"/>
  <title>self-signed.badssl.com</title>
  <link rel="stylesheet" href="/style.css">
  <style>body { background: red; }</style>
</head>
<body>
<div id="content">
  <h1 style="font-size: 12vw;">
    self-signed.<br>badssl.com
  </h1>
</div>

</body>
</html>
```

Any chance this can be changed for the next release? kostya's code looks like a good start?

**Comment by ysbaddaden** [comment]
The SSL integration is indeed lacking quality, and should be tackled. I started a PR to improve TLS in HTTP::Server, to provide a strong SSL context by default. I'll try to look closer at TLS for HTTP::Client too, namely: verifying the peer certificate, the hostname, specifying a context, ...

I found some documentation about [hostname validation in OpenSSL](https://wiki.openssl.org/index.php/Hostname_validation). The latest OpenSSL 1.0.2 introduces a [X509_check_host](https://www.openssl.org/docs/man1.0.2/crypto/X509_check_host.html) function, but this isn't available in previous versions that are still widely deployed (eg: Ubuntu Trusty). Alternatively we can tell the SSL socket about the hostname beforehand, so verifying the certificate will verify the hostname at the same time.

Looking at Ruby, the SSL sockets [have a method](https://github.com/ruby/ruby/blob/32674b167bddc0d737c38f84722986b0f228b44b/ext/openssl/lib/openssl/ssl.rb#L305) HTTP clients must call to verify the hostname once connected ([example](https://github.com/httprb/http/blob/bb4d5e19e3ea47d0b4d39908ab52e72d8faa48cb/lib/http/timeout/null.rb#L38)). This requires to manually verify the hostname against the peer certificate... which is hard work (eg: try writing bindings for the `X509` struct).
